From 0719c0526372c9d710f48b21c18610d03253b5f5 Mon Sep 17 00:00:00 2001
From: linya14x <linx.yang@intel.com>
Date: Fri, 14 Nov 2025 18:41:08 +0800
Subject: [PATCH] media: ipu: Update firmware ABI version to
 1.2.1.20251028_225604

media: ipu7: Workaround for FW breaking ABI compatibility
commit-id: ab9874f2747fd6af6cee621b05341f18d6212437
commit-id: c3b8dc8ad76895b6a9f17e705442ca19de05e76c
Signed-off-by: Hao Yao <hao.yao@intel.com>
Signed-off-by: linya14x <linx.yang@intel.com>
---
 .../staging/media/ipu7/abi/ipu7_fw_boot_abi.h |  1 +
 .../staging/media/ipu7/abi/ipu7_fw_isys_abi.h | 29 +++++++++++++++++--
 .../staging/media/ipu7/abi/ipu7_fw_msg_abi.h  |  2 +-
 drivers/staging/media/ipu7/ipu7-fw-isys.c     | 23 ++++++++++++++-
 4 files changed, 50 insertions(+), 5 deletions(-)

diff --git a/drivers/staging/media/ipu7/abi/ipu7_fw_boot_abi.h b/drivers/staging/media/ipu7/abi/ipu7_fw_boot_abi.h
index a1519c4fe6..4ce304f54e 100644
--- a/drivers/staging/media/ipu7/abi/ipu7_fw_boot_abi.h
+++ b/drivers/staging/media/ipu7/abi/ipu7_fw_boot_abi.h
@@ -153,6 +153,7 @@ enum ia_gofo_boot_state {
 	IA_GOFO_FW_BOOT_STATE_CRIT_MPU_CONFIG_FAILURE = 0xdead1013U,
 	IA_GOFO_FW_BOOT_STATE_CRIT_SHARED_BUFFER_FAILURE = 0xdead1014U,
 	IA_GOFO_FW_BOOT_STATE_CRIT_CMEM_FAILURE = 0xdead1015U,
+	IA_GOFO_FW_BOOT_STATE_CRIT_SYSCOM_CONTEXT_FAILURE = 0xDEAD1016U,
 	IA_GOFO_FW_BOOT_STATE_SHUTDOWN_CMD = 0x57a7f001U,
 	IA_GOFO_FW_BOOT_STATE_SHUTDOWN_START = 0x57a7e200U,
 	IA_GOFO_FW_BOOT_STATE_INACTIVE = 0x57a7e300U,
diff --git a/drivers/staging/media/ipu7/abi/ipu7_fw_isys_abi.h b/drivers/staging/media/ipu7/abi/ipu7_fw_isys_abi.h
index 45db85eb13..171b69c8c8 100644
--- a/drivers/staging/media/ipu7/abi/ipu7_fw_isys_abi.h
+++ b/drivers/staging/media/ipu7/abi/ipu7_fw_isys_abi.h
@@ -47,7 +47,6 @@ enum ipu7_insys_resp_type {
 	IPU_INSYS_RESP_TYPE_FRAME_EOF = 8,
 	IPU_INSYS_RESP_TYPE_STREAM_START_AND_CAPTURE_DONE = 9,
 	IPU_INSYS_RESP_TYPE_STREAM_CAPTURE_DONE = 10,
-	IPU_INSYS_RESP_TYPE_PWM_IRQ = 11,
 	N_IPU_INSYS_RESP_TYPE
 };

@@ -201,7 +200,8 @@ enum ipu7_insys_mipi_dt_rename_mode {
 enum ipu7_insys_output_link_dest {
 	IPU_INSYS_OUTPUT_LINK_DEST_MEM = 0,
 	IPU_INSYS_OUTPUT_LINK_DEST_PSYS = 1,
-	IPU_INSYS_OUTPUT_LINK_DEST_IPU_EXTERNAL = 2
+	IPU_INSYS_OUTPUT_LINK_DEST_IPU_EXTERNAL = 2,
+	N_IPU_INSYS_OUTPUT_LINK_DEST
 };

 enum ipu7_insys_dpcm_type {
@@ -220,9 +220,12 @@ enum ipu7_insys_dpcm_predictor {

 enum ipu7_insys_send_queue_token_flag {
 	IPU_INSYS_SEND_QUEUE_TOKEN_FLAG_NONE = 0,
-	IPU_INSYS_SEND_QUEUE_TOKEN_FLAG_FLUSH_FORCE = 1
+	IPU_INSYS_SEND_QUEUE_TOKEN_FLAG_FLUSH_FORCE = 1,
+	N_IPU_INSYS_SEND_QUEUE_TOKEN_FLAG
 };

+#define IPU_INSYS_MIPI_FRAME_NUMBER_DONT_CARE UINT16_MAX
+
 #pragma pack(push, 1)
 struct ipu7_insys_resolution {
 	u32 width;
@@ -349,6 +352,26 @@ struct ipu7_insys_buffset {
 };

 struct ipu7_insys_resp {
+	u64 buf_id;
+	struct ipu7_insys_capture_output_pin_payload pin;
+	struct ia_gofo_msg_err error_info;
+	u32 timestamp[2];
+	u16 mipi_fn;
+	u8 type;
+	u8 msg_link_streaming_mode;
+	u8 stream_id;
+	u8 pin_id;
+	u8 frame_id;
+	u8 skip_frame;
+};
+
+/**
+ * TODO: Revert this change after firmware ABI fixed.
+ * This is an internal workaround.
+ * IPU7.5 firmware changed its ABI and breaks
+ * compatibility with IPU7 Lunar Lake.
+ */
+struct ipu7_insys_legacy_resp {
 	u64 buf_id;
 	struct ipu7_insys_capture_output_pin_payload pin;
 	struct ia_gofo_msg_err error_info;
diff --git a/drivers/staging/media/ipu7/abi/ipu7_fw_msg_abi.h b/drivers/staging/media/ipu7/abi/ipu7_fw_msg_abi.h
index 8a78dd0936..1319f0eb63 100644
--- a/drivers/staging/media/ipu7/abi/ipu7_fw_msg_abi.h
+++ b/drivers/staging/media/ipu7/abi/ipu7_fw_msg_abi.h
@@ -217,7 +217,7 @@ struct ipu7_msg_task {
 	u8 frag_id;
 	u8 req_done_msg;
 	u8 req_done_irq;
-	u8 reserved[1];
+	u8 disable_save;
 	ipu7_msg_teb_t payload_reuse_bm;
 	ia_gofo_addr_t term_buffers[IPU_MSG_MAX_NODE_TERMS];
 };
diff --git a/drivers/staging/media/ipu7/ipu7-fw-isys.c b/drivers/staging/media/ipu7/ipu7-fw-isys.c
index 2958e39a35..26704c6fad 100644
--- a/drivers/staging/media/ipu7/ipu7-fw-isys.c
+++ b/drivers/staging/media/ipu7/ipu7-fw-isys.c
@@ -195,9 +195,30 @@ int ipu7_fw_isys_close(struct ipu7_isys *isys)

 struct ipu7_insys_resp *ipu7_fw_isys_get_resp(struct ipu7_isys *isys)
 {
-	return (struct ipu7_insys_resp *)
+	/**
+	 * TODO: Revert this change after firmware ABI fixed.
+	 * This is an internal workaround.
+	 * IPU7.5 firmware changed its ABI and breaks
+	 * compatibility with IPU7 Lunar Lake.
+	 */
+	struct ipu7_insys_resp *resp =
 		ipu7_syscom_get_token(isys->adev->syscom,
 				      IPU_INSYS_OUTPUT_MSG_QUEUE);
+
+	if (resp && isys->adev->isp->hw_ver == IPU_VER_7) {
+		struct ipu7_insys_legacy_resp lnl_resp;
+
+		memcpy(&lnl_resp, resp, sizeof(struct ipu7_insys_resp));
+		resp->mipi_fn = 0;
+		resp->type = lnl_resp.type;
+		resp->msg_link_streaming_mode = lnl_resp.msg_link_streaming_mode;
+		resp->stream_id = lnl_resp.stream_id;
+		resp->pin_id = lnl_resp.pin_id;
+		resp->frame_id = lnl_resp.frame_id;
+		resp->skip_frame = lnl_resp.skip_frame;
+	}
+
+	return resp;
 }

 void ipu7_fw_isys_put_resp(struct ipu7_isys *isys)
--
2.43.0

